util/oidc/provider.go

```diff
@@ -149,6 +162,126 @@ func (p *providerImpl) Verify(tokenString string, argoSettings *settings.ArgoCDS
 	return idToken, nil
 }
 
+// VerifyJWT verifies a JWT token using the configured JWK Set URL
+func (p *providerImpl) VerifyJWT(tokenString string, argoSettings *settings.ArgoCDSettings) (*jwt.Token, error) {
+	if argoSettings.JWTConfig == nil || argoSettings.JWTConfig.JWKSetURL == "" {
+		return nil, fmt.Errorf("JWT configuration not found")
+	}
+
+	cacheTTL := p.defaultCacheTTL
+	if argoSettings.JWTConfig.CacheTTL != "" {
+		ttl, err := time.ParseDuration(argoSettings.JWTConfig.CacheTTL)
+		if err != nil {
+			log.Warnf("Invalid JWT cache TTL %q, using default", argoSettings.JWTConfig.CacheTTL)
+		} else {
+			cacheTTL = ttl
+		}
+	}
+
+	jwks, err := p.getJWKS(argoSettings.JWTConfig.JWKSetURL, cacheTTL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get JWKS: %w", err)
+	}
+
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+
+		kid, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, fmt.Errorf("kid header not found in token")
+		}
+
+		var key *jose.JSONWebKey
+		for _, k := range jwks.Keys {
+			if k.KeyID == kid {
+				key = &k
+				break
+			}
+		}
+		if key == nil {
+			return nil, fmt.Errorf("no key found for kid %q", kid)
+		}
+
+		if key.Algorithm != "" && key.Algorithm != token.Header["alg"] {
+			return nil, fmt.Errorf("algorithm mismatch for kid %q: expected %v, got %v", kid, key.Algorithm, token.Header["alg"])
+		}
+
+		return key.Key, nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse/verify JWT: %w", err)
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, fmt.Errorf("invalid token claims")
+	}
+
+	if argoSettings.JWTConfig.EmailClaim != "" {
+		if _, ok := claims[argoSettings.JWTConfig.EmailClaim]; !ok {
+			return nil, fmt.Errorf("required email claim %q not found", argoSettings.JWTConfig.EmailClaim)
+		}
+	}
+
+	if argoSettings.JWTConfig.UsernameClaim != "" {
+		if _, ok := claims[argoSettings.JWTConfig.UsernameClaim]; !ok {
+			return nil, fmt.Errorf("required username claim %q not found", argoSettings.JWTConfig.UsernameClaim)
+		}
+	}
+
+	// Verify audience if configured
+	if aud, ok := claims["aud"].(string); ok {
+		if argoSettings.JWTConfig.Audience != "" && aud != argoSettings.JWTConfig.Audience {
+			return nil, fmt.Errorf("invalid audience claim, got %q, want %q", aud, argoSettings.JWTConfig.Audience)
```

Comment:
You used `expected` earlier -- is there a meaningful distinction between `want` and `expected`?

Reply:
In this case, yes. The audience is a set definition from ArgoCD that is not defined by the token. Basically an audience mismatch suggests either someone had compromised a token for another application. Above where I used `expected` it's where the token itself defined a different algorithm than the one it actually used, suggesting the token minting process went wrong. 

Small difference, and it isn't technically different, but they imply two different things.

Reply:
I've generally seen `got`/`want` in tests as opposed to user facing stuff. 

If you're trying to highlight something spooky, I think you should include more words as opposed to this minor wording difference. Keep in mind that a user will only see one error at a time and thus won't be aware of the alternate words.